salt/states/file.py

```diff
@@ -352,6 +352,33 @@ def _check_user(user, group):
     return err
 
 
+def _is_valid_relpath(
+        relpath,
+        maxdepth=None):
+    '''
+    Performs basic sanity checks on a relative path.
+
+    Requires slash-separated path components (i.e. the kind obtained
+    through cp.list_master or other such calls).
+
+    Ensures that the path does not contain directory transversal, and
+    that it does not exceed a stated maximum depth (if specified).
+    '''
+    # Check relpath surrounded by slashes, so that `..` can be caught as
+    # a path component at the start, end, and in the middle of the path.
+    if '/../' in '/{0}/'.format(relpath):
```

Comment:
I did end up looking into the differing relpath implementations though, and there is definitely some cause for concern. I'm considering using `posixpath` to deal with the salt master paths, and that just leaves the concern of what to do with the paths when actually copying to the filesystem (i.e. whether or not they'll need transformation). I'll give that an attempt today.